Refactor Deadline callbacks to use the new Callback model

ramitkataria opened the pull request:
<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
Followup to https://github.com/apache/airflow/pull/54796. This PR essentially moves callback functionality from deadline callbacks to the generic callbacks and refactors Deadline to use these callbacks.

cc @ferruzzi @jason810496 @potiuk @seanghaeli

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

ferruzzi commented on airflow-core/src/airflow/migrations/versions/0092_3_2_0_replace_deadline_inline_callback_with_fkey.py:
```diff
@@ -0,0 +1,82 @@
+    """Replace Deadline table's inline callback fields with callback_id foreign key."""
+    # Delete all rows for Deadline Alerts.
+    op.execute("DELETE FROM deadline")
+    # TODO: Implement data-preserving migration.
```
I feel like extracting the existing data should be part of this PR, but if you are confident you'll get it done before next release I guess I can defer.

ramitkataria replied:
I wanted to do it in a separate PR so that I don't have to do several iterations of testing the migration if there are changes requested in this PR that would affect the script. I would like the changes in this PR to be confirmed so that I can do thorough testing (with different cases of what the tables could contain) of the script only once.

The next release is expected for March 2026 so it should definitely be done by then.

ferruzzi replied:
I am working on a data migration myself right now and have changed my mind, this isn't as trivial as I had hoped.  I withdraw this comment and apologize.

ferruzzi replied:
Actually, going to leave this unresolved for now so this doesn't get accidentally merged until we have a chance to talk.

o-nikolas replied:
Did this get resolved offline? I see it's marked as resolved again @ferruzzi? If so can the context be put here for posterity?

ferruzzi replied:
I resolved it when I added the comment to approve it.  he convinced me that this more modular/granular change makes more sense.  I do agree with you above that the existing data shouldn't be wiped until after it is migrated over though

ramitkataria replied:
Yes, I would like to do it separately so that I don't have to do manual testing of the migration script in each iteration of the PR in case there are comments that require making changes to the schema. The migration script doesn't seem to be automatically tested with real data in the database

ramitkataria replied:
After offline discussion, I ended up adding the data-preserving migration in this PR

ferruzzi commented on airflow-core/src/airflow/models/callback.py:
```diff
@@ -85,7 +90,7 @@ class ImportPathCallbackDefProtocol(CallbackDefinitionProtocol, Protocol):
 
     path: str
-    kwargs: dict | None
+    kwargs: dict
```
Why are kwargs mandatory now?

ramitkataria replied:
See https://github.com/apache/airflow/pull/57215#discussion_r2467524125 but basically, kwargs are stored as `{}` when the user doesn't include them

ferruzzi commented on airflow-core/src/airflow/models/callback.py:
```diff
@@ -131,26 +136,42 @@ class Callback(Base):
     trigger = relationship("Trigger", back_populates="callback", uselist=False)
 
-    def __init__(self, priority_weight: int = 1):
+    def __init__(self, priority_weight: int = 1, prefix: str = "", **kwargs):
```
I think this needs a docstring.  `priority_weight` was fairly self-explanatory, but the purpose of `prefix` isn't.  I think the docstring on the __init__ explaining that and that kwargs get passed to the metric tags would help.

ramitkataria replied:
Just added the docstrings!

ferruzzi commented on airflow-core/src/airflow/models/callback.py:
```diff
@@ -131,26 +136,42 @@ class Callback(Base):
+    def __init__(self, priority_weight: int = 1, prefix: str = "", **kwargs):
         self.state = CallbackState.PENDING
         self.priority_weight = priority_weight
+        self.data = kwargs  # kwargs can be used to include additional info in metric tags
```
Unless there are other uses I haven't gotten to yet, maybe rename `data` to `metric_tags`?

ramitkataria replied:
`data` is used by the subclasses to store additional metadata that is not covered by the columns in the base class. This would include things like `path` and `kwargs`. It also includes `prefix`

ferruzzi commented on airflow-core/src/airflow/models/callback.py:
```diff
@@ -131,26 +136,42 @@ class Callback(Base):
+        tags = {"result": result, **self.data}
+        tags.pop("prefix", None)
+
+        if "kwargs" in tags:
+            # Remove the context (if exists) to keep the tags simple
+            tags["kwargs"] = {k: v for k, v in tags["kwargs"].items() if k != "context"}
```
Do we need a new dict for some reason?  If not, why not just do 

```
if "kwargs" in tags:
    tags["kwargs"].pop("context", None)` 
```

and skip the whole dict comprehension?

ramitkataria replied:
This is because of the shallow copy of the dict. If I just did a `pop`, it would also remove context from the kwargs that would be passed to the trigger at callback runtime

ferruzzi commented on airflow-core/src/airflow/models/callback.py:
```diff
@@ -131,26 +136,42 @@ class Callback(Base):
+            # Remove the context (if exists) to keep the tags simple
+            tags["kwargs"] = {k: v for k, v in tags["kwargs"].items() if k != "context"}
+
+        prefix = self.data.get("prefix", "")
+        name = f"{prefix}.callback_{status}" if prefix else f"callback_{status}"
```
I think you want to add the leading `.` to the prefix if prefix is not "", right?  otherwise without a prefix, all these metrics will start with a period

ramitkataria replied:
The way it is right now, I've assumed that you would always want a `.` when a prefix is included so I've made it so that you don't need to include the `.` in the prefix parameter. If no prefix is included, the `.` is also not included (in the `else` case). However, I'm not too opinionated about this and would be happy to change if you prefer that the caller include the `.`

ferruzzi replied:
I totally misread that, I'm sorry.

ferruzzi commented on task-sdk/src/airflow/sdk/definitions/deadline.py:
```diff
@@ -121,13 +121,13 @@ class Callback(ABC):
 
     path: str
-    kwargs: dict | None
+    kwargs: dict
```
Same as above, why isn't this optional anymore?

ramitkataria replied:
kwargs is still an optional parameter in the constructor. I've just changed the internal representation so that null kwargs are stored as an empty dict. This helps with reducing `if` statements and edge cases

ferruzzi commented on airflow-core/tests/unit/models/test_callback.py:
```diff
@@ -72,7 +76,7 @@ def test_create_from_sdk_def(self, callback_def, expected_cb_instance):
         assert isinstance(returned_cb, type(expected_cb_instance))
         assert returned_cb.data == expected_cb_instance.data
 
-    def test_test_create_from_sdk_def_unknown_type(self):
```
heh.  oops :sweat_smile:

ramitkataria replied:
😅

ferruzzi reviewed (commented):
Left some questions.

ferruzzi reviewed (approved):
Had a chat in person and I'm convinced.  Resolved my concerns and approving.

o-nikolas commented on airflow-core/src/airflow/migrations/versions/0092_3_2_0_replace_deadline_inline_callback_with_fkey.py:
```diff
@@ -0,0 +1,82 @@
+    # Delete all rows for Deadline Alerts.
+    op.execute("DELETE FROM deadline")
+    # TODO: Implement data-preserving migration.
+    # Deadline Alerts was an experimental feature in 3.1 so deleting past data isn't a breaking change anyway.
```
Experimental usually means that the API can change in backwards incompatible ways. Stark data loss isn't usually expected by users.

ramitkataria replied:
I see, I'll remove that code comment but I'm planning to make a separate PR for that anyway

ramitkataria committed:
Refactor Deadline callbacks to use the new Callback model

Followup to #54796. This PR essentially moves callback functionality
from deadline callbacks to the generic callbacks and refactors Deadline
to use these callbacks.

# Conflicts:
#	airflow-core/src/airflow/migrations/versions/0091_3_2_0_restructure_callback_table.py

# Conflicts:
#	airflow-core/src/airflow/models/deadline.py

ramitkataria committed:
Implement DB migration

# Conflicts:
#	airflow-core/docs/img/airflow_erd.sha256
#	airflow-core/docs/img/airflow_erd.svg

amoghrajesh commented on airflow-core/src/airflow/models/callback.py:
```diff
@@ -131,26 +136,49 @@ class Callback(Base):
     def queue(self):
         self.state = CallbackState.QUEUED
 
+    def get_metric_info(self, status: str, result: Any) -> dict:
```
Post merge comment: Can we make the `status` take a `CallbackState` object instead? This fails the tests in 3.11+ python and the reason for that is enum -> str interpolation changes in python3.11+, as an example see this:

3.10:

```python
Python 3.10.15 (main, Oct  8 2024, 00:44:52) [Clang 18.1.8 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from enum import Enum
>>> 
>>> class CallbackState(str, Enum):
...     SUCCESS = "success"
...     FAILURE = "failure"
... 
>>> status = CallbackState.SUCCESS
>>> 
>>> print("str(status):", str(status))
str(status): CallbackState.SUCCESS
>>> print("f-string:", f"{status}")
f-string: success
>>> print("Using .value:", f"{status.value}")
Using .value: success

```

3.13:
```python
(airflow) ➜  airflow git:(callback-tests) ✗ python                                                         
Python 3.13.3 (main, Apr  8 2025, 13:54:08) [Clang 17.0.0 (clang-1700.0.13.3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from enum import Enum
... 
... class CallbackState(str, Enum):
...     SUCCESS = "success"
...     FAILURE = "failure"
... 
... status = CallbackState.SUCCESS
... 
... print("str(status):", str(status))
... print("f-string:", f"{status}")
... print("Using .value:", f"{status.value}")
... 
str(status): CallbackState.SUCCESS
f-string: CallbackState.SUCCESS
Using .value: success
```

amoghrajesh replied:
I have fixed it for the meanwhile this way: https://github.com/apache/airflow/pull/57935

ramitkataria replied:
Thanks! Fixed in https://github.com/apache/airflow/pull/58168

amoghrajesh reviewed (commented):
@ramitkataria just a post merge commit
